### PR TITLE
refactor(core): cleanup prefetch triggers when resource loading starts

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit adds the necessary mechanisms to perform cleanup of prefetch triggers when resource loading starts. Previously, this logic was missing, which resulted in retaining those triggers.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/instructions/defer.ts [hunk 1]
```diff
@@ -693,3 +702,71 @@ export function getDeferBlocks(lView: LView, deferBlocks: DeferBlockDetails[]) {
...
+    for (const [block] of this.blocks) {
+      this.cleanup(block);
+    }
+    this.blocks.clear();
+  }
+
+  /** @nocollapse */
+  static ɵprov = /** @pureOrBreakMyCode */ ɵɵdefineInjectable({
+    token: DeferBlockCleanupManager,
+    providedIn: 'root',
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/defer.ts:769]: Could we provide this even higher, considering that it's not injecting anything? E.g. `environment`

File: packages/core/src/render3/instructions/defer.ts [hunk 2]
```diff
@@ -693,3 +702,71 @@ export function getDeferBlocks(lView: LView, deferBlocks: DeferBlockDetails[]) {
...
+    const block = this.blocks.get(tDetails)!;
+    if (!block.has(key)) {
+      block.set(key, []);
+    }
+    const callbacks = block.get(key)!;
+    callbacks.push(callback);
+  }
+
+  has(tDetails: TDeferBlockDetails, key: string): boolean {
+    return this.blocks.has(tDetails) && this.blocks.get(tDetails)!.has(key);
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/defer.ts:744]: Could shorten this to `!!this.blocks.get(tDetails)?.has(key)`

File: packages/core/src/render3/interfaces/defer.ts [hunk 3]
```diff
@@ -14,16 +14,25 @@ import type {DependencyType} from './definition';
  */
 export type DependencyResolverFn = () => Array<Promise<DependencyType>>;
 
+/**
+ * Enumerates all `on` triggers of a defer block.
+ */
+export enum DeferBlockTriggers {
```
Comment (crisbeto) [inline packages/core/src/render3/interfaces/defer.ts:20]: Do we plan to also use this for the non-prefetch triggers? If not, we could just inline a string when calculating the key since it's only used in one place.

**Review by crisbeto — APPROVED** [review]

File: packages/core/src/render3/interfaces/defer.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/interfaces/defer.ts:20]: Ah, I meant to do `const enum`, so that the value is inlined (so we get the benefit of having all triggers in one place without extra cost).

File: packages/core/src/render3/instructions/defer.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/defer.ts:769]: My idea was that we'd also store cleanup fns for interaction, hover and viewport there and those cleanup fns would close over DOM elements. In this case, when an application is removed and `ngOnDestroy` is called, we also invoke cleanup fns to remove listeners and release DOM references.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:567521549:BASE:567521569:1695361603812:f41ab0c6).

File: packages/core/src/render3/instructions/defer.ts (hunk 1 above)
Comment (crisbeto) [reply packages/core/src/render3/instructions/defer.ts:769]: I don't think that would be necessary since the other `on` triggers register themselves with the cleanup function of the `LView` they're on so they'll be removed when the app is destroyed.

File: packages/core/src/render3/interfaces/defer.ts (hunk 3 above)
Comment (crisbeto) [reply packages/core/src/render3/interfaces/defer.ts]: Worth keeping this change in mind https://github.com/angular/angular/pull/51670.

File: packages/core/src/render3/instructions/defer.ts (hunk 1 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/defer.ts:769]: Discussed offline: the cleanup manager would be a centralized place to coordinate cleanup for all prefetch triggers (we'd probably need to rename the class to `PrefetchTriggerManager` in a followup).

**Comment by AndrewKushnir** [comment]
**Caretaker note**: the `framework-win` CI failure looks unrelated and pre-existing. Presubmit is "green". This PR should be ready for merge.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 917203d9916096ea237f5a4329052f01e0a9dbf9.
